src/cider/nrepl/middleware/debug.clj

```diff
@@ -344,6 +349,111 @@
+                         (read []
+                           (let [i (proxy-super read)]
+                             (when @collect?
+                               (.append text (char i)))
+                             i))
+                         (unread [c]
+                           (when @collect?
+                             (.deleteCharAt text (dec (.length text))))
+                           (proxy-super unread c)))
+              ;; Fix bogus column number of 1, which really means 0
+              column   (if (= 1 column) 0 column)]
+
+          ;; Give it the right line and column number. We can just set the
+          ;; line number directly, but there is no setColumnNumber method, so
+          ;; we have to hack it a bit.
+          (.setLineNumber pbr (.getLineNumber rdr))
+          (dotimes [_ column]
+            (.read pbr))
+
+          ;; Now start collecting the code
+          (reset! collect? true)
+
+          (let [form (read {} pbr)
+                code (str text)]
+            (assoc var-meta
+                   :form form
+                   :code code)))))))
+
+(declare debug-reader)
+
+(defn instrument-var-for-step-in
+  "Attach an instrumented version of the function in `v` as metadata to `v`,
+  leaving the contents of the var uninstrumented."
+  [v]
+  (when-not (::instrumented (meta v))
+    (when-let [{:keys [ns file form] :as var-meta} (read-var v)]
+      (let [full-path (misc/transform-value (:file (info/file-info file)))]
+        (binding [*ns*   ns
+                  *file* file
+                  *msg*  (-> *msg*
+                             (merge var-meta)
+                             (assoc :file full-path))]
+          (-> form
+              debug-reader
+              ins/instrument-tagged-code
+              eval)
+          (let [instrumented @v]
+            (eval form)
+            (alter-meta! v assoc ::instrumented instrumented)))))))
+
+(defn step-in?
+  "Return true if we can and should step in to the function in the var `v`.
+  The \"should\" part is determined by the value in `step-in-to-next?`, which
+  gets set to true by the user sending the \"step in\" command."
+  [v coor]
+  (when (and @step-in-to-next?
+             (not (skip-breaks? coor))
+             (not= :trace (:mode @*skip-breaks*))
+             (not (::ins/breakfunction (meta v))))
+    (try
+      (instrument-var-for-step-in v)
+      true
+      (catch Exception e
+        ;; TODO - how do we inform the user that we failed?
+        false))))
+
+(defn looks-step-innable?
+  "Decide whether a form looks like a call to a function that we could
+  instrument and step into."
+  [form]
+  (when (and (seq? form) (symbol? (first form)))
+    (let [v (resolve (first form))
+          m (meta v)]
+      ;; We do not go so far as to actually try to read the code of the function
+      ;; at this point, which is at macroexpansion time.
+      (and v
+           (not (:macro m))
```

Comment:
Might also want to check if it's not a special form. The full list is:

    if do recur throw finally try monitor-exit monitor-enter new quote & var clojure.core/import* . def set! loop* let* letfn* reify* deftype* fn* catch case*